src/subtype.c

```diff
@@ -3014,6 +3014,39 @@ static jl_value_t *intersect_unionall_(jl_value_t *t, jl_unionall_t *u, jl_stenv
     return res;
 }
 
+static int always_occurs_cov(jl_value_t *v, jl_tvar_t *var, int param) JL_NOTSAFEPOINT
+{
+    if (param > 1) {
+        return 0;
+    }
+    else if (v == (jl_value_t*)var) {
+        return param == 1;
+    }
+    else if (jl_is_uniontype(v)) {
+        return always_occurs_cov(((jl_uniontype_t*)v)->a, var, param) ||
+               always_occurs_cov(((jl_uniontype_t*)v)->b, var, param);
+    }
+    else if (jl_is_unionall(v)) {
+        jl_unionall_t *ua = (jl_unionall_t*)v;
+        return ua->var != var && (
+            always_occurs_cov(ua->var->lb, var, 0) ||
+            always_occurs_cov(ua->var->ub, var, 0) ||
```

Comment:
```suggestion
            always_occurs_cov(ua->var->lb, var, 1) ||
            always_occurs_cov(ua->var->ub, var, 0) ||
```

To make this valid, it might also need to track whether the var here ends up eventually constraining it?
```
&& !occurs(ua->var->ub, var) || always_occurs_cov(ua->body, ua->var, param))
```

Reply:
I think I misunderstood the meaning of the `always_occurs_cov` to mean `only_cov`, which was not the intent. I guess this function is actually equivalent to `Core.Compiler.constrains_param && !occurs_inv`

A use as a lower bound does not create a constraint on the lower bound for var:
```suggestion
            always_occurs_cov(ua->var->ub, var, 0) ||
```

Reply:
I guess they have some little difference? But the suggestions make sense (As their task looks similar).

`always_occurs_cov` should perform a conservative estimation on the `occurs_cov` after 1st intersection.
As we want to match the branch `if (vb.occurs_cov && noinv)` below.

`var_occurs_inside` is not a good template here:
1. We only perform `intersect_aside` on `ub`.
2. The union decision should not affect the conservativeness.